### ISSUE TITLE
Release 7.11

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#2573]
- [x] stylelint release
- [x] stylelint-config-standard update/release
- [x] stylelint-demo update
- [x] stylelint.io update
- [x] tweet

In https://github.com/stylelint/stylelint/issues/2532 it feels like we've mapped a way forward. One that I believe will get us both get back to regular releases and bring `8.0.0` closer.

As per https://github.com/stylelint/stylelint/issues/2532#issuecomment-301243045, we should revert (for now) https://github.com/stylelint/stylelint/pull/2500 until we are in a better position to assess the viability of such an approach. An approach that has to overcome the [limitations](https://github.com/stylelint/stylelint/issues/2532#issuecomment-301222916) of our design decision to use standalone rules.

Once this is done, I think we should release `7.11`.

**Comment by hudochenkov** [comment]
I don't see a reason to revert #2500. There is no pretty-printers, and there won't be any time soon. Because we need to develop support for them. And community should develop one. Rules in #2500 could already help people.

**Comment by jeddy3** [comment]
For the reasons laid out in https://github.com/stylelint/stylelint/issues/2532#issuecomment-301243045 & https://github.com/stylelint/stylelint/issues/2532#issuecomment-301222916, I'm not sure we've in a good position to know whether opening that can of worms is going to be good for the long term health of the project.

**Comment by hudochenkov** [comment]
> I agree that configuration-based whitespace formatting would give people more control. However, I propose we prioritise adding --fix support to the non-whitespace stylistic rules as this would benefit both camps.

I thought the point was to prioritise, not _ban_ whitespace related rules.

Without autofixing for whitespace related rules users have very few choices:

* If there is no opioinated pretty-printers for their liking, they have to fix everything manually as it is now.
* If there is pretty-printer for their liking, but few changes drives them crazy. They could only not use this pretty-printer and do everything manually again.

With autofixing for whitespace rules users can:

* Use only them for formatting, if they prefer.
* Use pretty-printer and override some moments with rule autofixing.

If we ban whitespace related rules, then we postpone again fixing most commons problems for most users. I believe most commons problems are whitespace problems.

**Comment by jeddy3** [comment]
I'm suggesting we postpone starting down the path of trying to autofix whitespace rules until we're in a better position to answers at the end of https://github.com/stylelint/stylelint/issues/2532#issuecomment-301243045:

- How many of the remaining users would benefit from adding `--fix` to stylelint's whitespace rules?
- How much additional code, and of what complexity, is needed to cater for this use case?
- How much is the sustainability of the project affected by this additional complexity?

I believe the golden path is:

1. stylelint autofixes non-whitespace stylistic aspects.
    - This is compatible with our [our original design decision](https://github.com/stylelint/stylelint/issues/1#issuecomment-68450190) to build standalone rules, whereas trying to autofix whitespace _fights against_ this design by introducing transformations whom's side-effects can fail other rules that have already been checked.
2. A prettifier algorithmically formats all the whitespace _and gets it 100% right_.
    - This is about a team accepting the opinionated logic of the prettifier. I hope that's what we intend to do in https://github.com/stylelint/stylelint/issues/2518, otherwise, like I explain in https://github.com/stylelint/stylelint/issues/2532#issuecomment-301243045 using https://github.com/stylelint/eslint-config-stylelint/issues/22 as an example, we're back to square one.

I suggesting we postpone trying to cater for the use cases you describe above until we gone down this golden path. Only then will we know _what proportion_ of the user base isn't catered for by the golden path, and only then we'll be able to adequately answer the questions above.

In short, for me, this issue is about balancing short-term gains against longer term sustainability of the project, while meeting the needs of the _majority_ of users.

**Comment by hudochenkov** [comment]
> How much additional code, and of what complexity, is needed to cater for this use case?

We can see it already, because we have 6 rules with autofixing, and 2 PRs waiting.

---

How many rules we'll ship in 7.11 then? Right now we only have one non-whitespace rule merged, and one rule in pending PR. I think we need a little more.

In which order stylelint runs rules, when plugins enabled? Core rules, than plugins rules?

When `--prettier` work begin? And who would like to do this?

**Comment by jeddy3** [comment]
> In which order stylelint runs rules, when plugins enabled? Core rules, than plugins rules?

This gets to the crux of the issue. If we only fix non-whitespace stylistic issues than I believe _it doesn't matter what order rules or plugins are run in_. This is why I believe my proposal is compatible with the original design decision of standalone rules. _No further work is needed to resolve side effects_.

> We can see it already, because we have 6 rules with autofixing, and 2 PRs waiting.

Yes, I agree we can make an assessment on complexity now. But we can't yet judge if this is the primary use case or not, and, in turn, we can't make an assessment on how to balance this complexity with the long-term sustainability of the project.

> How many rules we'll ship in 7.11 then?

It is OK if we just ship with just one. I think it's more important we release bug fixes in `head`.

> When --prettier work begin? 

It's not tied into any other features or release, and so work can start on it whenever. 

> And who would like to do this?

https://github.com/stylelint/stylelint/issues/2532 is stylelint's most starred issue ever. Hopefully, this equates to will within user base to assist building it.

**Comment by hudochenkov** [comment]
Would you mind update this issue with new plan discovered in https://github.com/stylelint/stylelint/issues/2532, please?

**Comment by hudochenkov** [comment]
I think we could postpone [`rule.dependsOn`](https://github.com/stylelint/stylelint/issues/2532#issuecomment-302109499) feature until 7.12. Because autofixing is in good shape for the first release, and making `dependsOn` quick, but [“dumb”](https://github.com/stylelint/stylelint/issues/2532#issuecomment-302140569) may not worth effort and 7.11 delay.

There few PRs I think we should merge for 7.11 release:

- [ ] [Indentation autofixing](https://github.com/stylelint/stylelint/pull/2529)
- [ ] [Add autofixing for color-hex-case](https://github.com/stylelint/stylelint/pull/2577)
- [ ] [Fixed: ignore variables with signs in `value-keyword-case`](https://github.com/stylelint/stylelint/pull/2558)
- [ ] [Update resolve-from to the latest version](https://github.com/stylelint/stylelint/pull/2539)

**Comment by jeddy3** [comment]
Is `--fix` for the `*-empty-line-before` rules dependent on the `indentation` rule always running _afterwards_? Or does it not matter for these rules?

**Comment by hudochenkov** [comment]
It's not depends, but recommended to run for better results. The only situation when `indentation`  is better be run after `*-empty-line-before` rules when node for `*-empty-line-before` rule is not on it's own line.

Given:

```json
{
    "declaration-empty-line-before": true
}
```

Before:

```css
div {
    color: red; display: inline-block;
}
div {
    color: red;
    display: inline-block;
}
```

After:

```css
div {
    color: red;

 display: inline-block;
}
div {
    color: red;

    display: inline-block;
}
```

Given:

```json
{
    "declaration-empty-line-before": true,
    "indentation": 4
}
```

Before:

```css
div {
    color: red; display: inline-block;
}
div {
    color: red;
    display: inline-block;
}
```

After:

```css
div {
    color: red;

    display: inline-block;
}
div {
    color: red;

    display: inline-block;
}
```

**Comment by hudochenkov** [comment]
As a temporary solution `indentation` is [placed](https://github.com/stylelint/stylelint/pull/2529/files#diff-1b5afd6b670c8bdd40102c599cbc9697) as the very last rule.

**Comment by jeddy3** [comment]
> The only situation when indentation is better be run after *-empty-line-before rules when node for *-empty-line-before rule is not on it's own line.

Thanks for the examples. It's clearer to me now.

> As a temporary solution indentation is placed as the very last rule.

I missed that somehow! Thanks for pointing it out.

> There few PRs I think we should merge for 7.11 release:

Agreed. Let's get those in and release.

**Comment by hudochenkov** [comment]
One more issue is important to be resolved for 7.11 https://github.com/stylelint/stylelint/issues/2607

**Comment by hudochenkov** [comment]
I have a good news as well. I've been working on the new `stylelint-order` version, and it has autofixing! Will publish it as soon as 7.11 released.

I've tried stylelint `master` branch and new `stylelint-order` on two big projects and they work very well. They fix what they can fix, and don't add new errors :)

**Comment by hudochenkov** [comment]
Only one issue left to solve for 7.11 release! But it's a tricky one #2607

**Comment by hudochenkov** [comment]
Well. Everything is merged and fixed for this release. @stylelint/core shall we make a release?

**Comment by evilebottnawi** [comment]
@hudochenkov yep

**Comment by jeddy3** [comment]
> Well. Everything is merged and fixed for this release.

Fantastic! 

> shall we make a release?

Yep. I can cut a release later today. 

(Reminder to self to document the release process!)

**Comment by jeddy3** [comment]
🎉 Released 🎉 

Thanks everyone for your contributions!

**Comment by m-allanson** [comment]
🚂🚋🚋🚋💨 Version 8 here we come!

**Comment by hudochenkov** [comment]
Yeah! I'm so happy we've released it :)

@m-allanson there is also [7.12](https://github.com/stylelint/stylelint/issues/2574) :)
